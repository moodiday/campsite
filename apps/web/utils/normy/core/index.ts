export type { NormalizerConfig, Data } from './types'
export { createNormalizer } from './create-normalizer'
export { getId } from './get-id'
