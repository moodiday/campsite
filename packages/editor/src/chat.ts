import { NodeViewRenderer } from '@tiptap/core'

import * as E from './extensions'

export interface GetChatExtensionsOptions {
  link?: Partial<E.LinkOptions>
  placeholder?: string
  mention?: Partial<E.MentionOptions>
  submit?: Partial<E.EnterSubmitOptions> & { enabled?: boolean }
  relativeTime?: Partial<E.RelativeTimeOptions> & {
    disabled?: boolean
    addNodeView?(): NodeViewRenderer
  }
}

export function getChatExtensions(options?: GetChatExtensionsOptions) {
  return [
    ...E.StarterKit(),

    E.Bold,
    E.Italic,
    E.Strike,
    // chat doesn't support underline so remove it from the exclude list
    E.CodeWithoutUnderline,
    E.Link.configure(options?.link),
    E.Typography,

    ...(options?.relativeTime?.disabled !== true
      ? [
          E.RelativeTime.extend({
            addNodeView: options?.relativeTime?.addNodeView
          }).configure(options?.relativeTime)
        ]
      : []),

    E.BlurOnEscape,
    E.Placeholder.configure({
      placeholder: options?.placeholder
    }),

    ...(options?.submit?.enabled ? [E.EnterSubmit.configure(options.submit)] : []),

    E.Mention.configure(options?.mention),
    E.Reaction
  ]
}
