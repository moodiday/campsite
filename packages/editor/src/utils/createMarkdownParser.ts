import { AnyExtension } from '@tiptap/core'
import { Schema } from '@tiptap/pm/model'
import MarkdownIt from 'markdown-it'
import Token from 'markdown-it/lib/token'

import markdownItTasks from './markdownItTasks'
import { MarkdownParser, ParseSpec } from './MarkdownParser'

const ignoreAllMarkdownTokens: { [name: string]: ParseSpec } = {
  // blocks
  paragraph: { block: 'paragraph' },
  blockquote: { block: 'paragraph' },
  code_block: { block: 'paragraph' },
  fence: { block: 'paragraph' },
  heading: { block: 'paragraph' },
  hr: { ignore: true, noCloseToken: true },
  ordered_list: { ignore: true },
  bullet_list: { ignore: true },
  list_item: { ignore: true },
  references: { block: 'paragraph' },
  table: { block: 'paragraph' },
  thead: { ignore: true },
  th: { ignore: true },
  tbody: { ignore: true },
  tr: { ignore: true },
  td: { ignore: true },
  image: { block: 'paragraph', noCloseToken: true },
  hardbreak: { ignore: true, noCloseToken: true },
  softbreak: { ignore: true, noCloseToken: true },

  // inline
  code_inline: { ignore: true },
  strong: { ignore: true },
  em: { ignore: true },
  link: { ignore: true },
  s: { ignore: true }
}

export function listIsTight(tokens: readonly Token[], i: number) {
  while (++i < tokens.length) if (tokens[i].type != 'list_item_open') return tokens[i].hidden
  return false
}

function getMarkdownSpec(extension: AnyExtension) {
  if ('markdownParseSpec' in extension.config === false || typeof extension.config.markdownParseSpec !== 'function') {
    return null
  }

  return extension.config.markdownParseSpec() as ParseSpec | null | undefined
}

function getMarkdownToken(extension: AnyExtension) {
  if ('markdownToken' in extension.config === false || typeof extension.config.markdownToken !== 'string') {
    return extension.name
  }
  return extension.config.markdownToken
}

export const createMarkdownParserSpec = (spec: ParseSpec) => spec

export function createMarkdownParser(
  schema: Schema,
  extensions: AnyExtension[],
  // WARNING: Provide JSDOM window and document objects to avoid errors in Node.js environment
  domParser: DOMParser = new window.DOMParser(),
  document: Document = window.document
) {
  const tokens = extensions
    .filter(
      (extension) => 'markdownParseSpec' in extension.config && typeof extension.config.markdownParseSpec === 'function'
    )
    .reduce((nodes, extension) => {
      const parseSpec = getMarkdownSpec(extension)

      if (!parseSpec) {
        return nodes
      }

      return {
        ...nodes,
        [getMarkdownToken(extension)]: parseSpec
      }
    }, ignoreAllMarkdownTokens)

  return new MarkdownParser(
    schema,
    MarkdownIt('default', {
      // enables raw inline html with the html_inline token
      html: true,
      breaks: false,
      linkify: true
    }).use(markdownItTasks),
    tokens,
    domParser,
    document
  )
}
