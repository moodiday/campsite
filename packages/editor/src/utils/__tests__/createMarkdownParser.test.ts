import { getSchemaByResolvedExtensions } from '@tiptap/core'
import { describe, expect, it } from 'vitest'

import { getChatExtensions } from '../../chat'
import { getMarkdownExtensions } from '../../markdown'
import { getNoteExtensions } from '../../note'
import { createMarkdownParser } from '../createMarkdownParser'

describe('createMarkdownParser', () => {
  const SAMPLE_MARKDOWN = `
# Heading 1
## Heading 2
### Heading 3
#### Heading 4
##### Heading 5
###### Heading 6

This is a paragraph with **bold** and italics.

- Bullet one
- Bullet two

1. Number one
2. Number two

\`\`\`
const foo = "bar"
\`\`\`

Hard break
Soft break

---

![CleanShot 2024-03-22 at 16 42 23@2x](https://github.com/campsite/campsite/assets/739696/49b398b1-8c03-4255-a759-21b8b53a3f5d)

| Header | Header | Header |
|--------|--------|--------|
| Cell | Cell | Cell |
| Cell | Cell | Cell | 

> Ullamco eiusmod laborum minim nulla adipisicing incididunt occaecat consequat non ipsum ex qui excepteur culpa.

And [here](https://linear.app/campsite/issue/CAM-6845/normalize-notes) is a link. With a \`inline code\` mark.
  `

  it('parses markdown extensions', () => {
    const extensions = getMarkdownExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(SAMPLE_MARKDOWN)

    expect(result).toMatchSnapshot()
  })

  it('parses note extensions', () => {
    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(SAMPLE_MARKDOWN)

    expect(result).toMatchSnapshot()
  })

  it('parses chat extensions', () => {
    const extensions = getChatExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(SAMPLE_MARKDOWN)

    expect(result).toMatchSnapshot()
  })

  it('converts softbreaks to hardbreaks with markdown', () => {
    const md = `
Foo
Bar
    `
    const extensions = getMarkdownExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })
  it('converts softbreaks to hardbreaks with note', () => {
    const md = `
Foo
Bar
    `
    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('ignores API mentions', () => {
    const md = `
Hey <@U123456> thanks for letting me know!
    `
    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('converts known HTML', () => {
    const md =
      'This is **bold** and this is a mention <span data-type="mention" data-id="abcdefabcdef" data-label="User Name" data-username="username">@User Name</span> and a member mention <span data-type="mention" data-id="abcdefabcdef" data-label="User Name" data-role="member" data-username="username">@User Name</span>.'

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('converts marks in HTML', () => {
    const md = 'This is <strong>bold</strong>.'

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('converts HTML blocks with markdown inside', () => {
    const md = `
<details>

## Heading

- List item
- List item

This is some *bold* text.

</details>
`

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('converts self-closing HTML', () => {
    const md =
      'Hey <span data-type="mention" data-id="abcdefabcdef" data-label="User Name" data-username="username" />.'

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('retains unknown inline HTML', () => {
    const md = `
Bad prompt:
<prompt>
"Help me with a presentation."
</prompt>
`

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('retains unknown block HTML', () => {
    const md = `
Bad prompt:

<prompt>
"Help me with a presentation."
</prompt>
`

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('handles ending with supported HTML', () => {
    const md = `
Foo bar baz

<span data-type="mention" data-id="abcdefabcdef" data-label="User Name" data-username="username" />
`

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('handles unsupported self-closing tag', () => {
    const md = `
What should we do with this <prompt />?
`

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('handles supported opening tag with no close', () => {
    const md = `
<p>
Foo bar baz
`

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('handles unsupported opening tag with no close', () => {
    const md = `
<prompt>
Foo bar baz
`

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('handles supported closing tag with no open', () => {
    const md = `
Foo bar baz
</p>
`

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('handles supported closing tag with no open in middle of paragraph', () => {
    const md = `
Foo bar baz

</p>

Cat dog fish
`

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('handles unsupported closing tag with no open', () => {
    const md = `
Foo bar baz
</prompt>
`

    const extensions = getNoteExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('handles task lists', () => {
    const md = `
- [ ] Task one
- [x] Task two
    `
    const extensions = getMarkdownExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })

  it('strips unsupported link attribute values', () => {
    const md =
      'This is <a href="https://campsite.com" rel="noopener noreferrer expect author preload" target="_parent" class="prose-link py-10 bg-red-500">link</a>.'

    const extensions = getMarkdownExtensions()
    const schema = getSchemaByResolvedExtensions(extensions)
    const parser = createMarkdownParser(schema, extensions)
    const result = parser.parse(md)

    expect(result).toMatchSnapshot()
  })
})
