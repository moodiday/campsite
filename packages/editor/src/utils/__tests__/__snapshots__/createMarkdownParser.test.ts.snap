// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createMarkdownParser > converts HTML blocks with markdown inside 1`] = `
{
  "content": [
    {
      "content": [
        {
          "content": [
            {
              "text": "Heading",
              "type": "text",
            },
          ],
          "type": "detailsSummary",
        },
        {
          "content": [
            {
              "content": [
                {
                  "content": [
                    {
                      "content": [
                        {
                          "text": "List item",
                          "type": "text",
                        },
                      ],
                      "type": "paragraph",
                    },
                  ],
                  "type": "listItem",
                },
                {
                  "content": [
                    {
                      "content": [
                        {
                          "text": "List item",
                          "type": "text",
                        },
                      ],
                      "type": "paragraph",
                    },
                  ],
                  "type": "listItem",
                },
              ],
              "type": "bulletList",
            },
            {
              "content": [
                {
                  "text": "This is some ",
                  "type": "text",
                },
                {
                  "marks": [
                    {
                      "type": "italic",
                    },
                  ],
                  "text": "bold",
                  "type": "text",
                },
                {
                  "text": " text.",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "detailsContent",
        },
      ],
      "type": "details",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > converts known HTML 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "This is ",
          "type": "text",
        },
        {
          "marks": [
            {
              "type": "bold",
            },
          ],
          "text": "bold",
          "type": "text",
        },
        {
          "text": " and this is a mention ",
          "type": "text",
        },
        {
          "attrs": {
            "id": "abcdefabcdef",
            "label": "User Name",
            "role": "member",
            "username": "username",
          },
          "type": "mention",
        },
        {
          "text": " and a member mention ",
          "type": "text",
        },
        {
          "attrs": {
            "id": "abcdefabcdef",
            "label": "User Name",
            "role": "member",
            "username": "username",
          },
          "type": "mention",
        },
        {
          "text": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > converts marks in HTML 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "This is ",
          "type": "text",
        },
        {
          "marks": [
            {
              "type": "bold",
            },
          ],
          "text": "bold",
          "type": "text",
        },
        {
          "text": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > converts self-closing HTML 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Hey ",
          "type": "text",
        },
        {
          "attrs": {
            "id": "abcdefabcdef",
            "label": "User Name",
            "role": "member",
            "username": "username",
          },
          "type": "mention",
        },
        {
          "text": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > converts softbreaks to hardbreaks with markdown 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Foo",
          "type": "text",
        },
        {
          "type": "hardBreak",
        },
        {
          "text": "Bar",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > converts softbreaks to hardbreaks with note 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Foo",
          "type": "text",
        },
        {
          "type": "hardBreak",
        },
        {
          "text": "Bar",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > handles ending with supported HTML 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Foo bar baz",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "attrs": {
            "id": "abcdefabcdef",
            "label": "User Name",
            "role": "member",
            "username": "username",
          },
          "type": "mention",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > handles supported closing tag with no open 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Foo bar baz",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "</p>
",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > handles supported closing tag with no open in middle of paragraph 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Foo bar baz",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "</p>
",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Cat dog fish",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > handles supported opening tag with no close 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Foo bar baz",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > handles task lists 1`] = `
{
  "content": [
    {
      "content": [
        {
          "attrs": {
            "checked": false,
          },
          "content": [
            {
              "content": [
                {
                  "text": "Task one",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "taskItem",
        },
        {
          "attrs": {
            "checked": true,
          },
          "content": [
            {
              "content": [
                {
                  "text": "Task two",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "taskItem",
        },
      ],
      "type": "taskList",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > handles unsupported closing tag with no open 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Foo bar baz",
          "type": "text",
        },
        {
          "type": "hardBreak",
        },
        {
          "text": "</prompt>",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > handles unsupported opening tag with no close 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "<prompt>
Foo bar baz
",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > handles unsupported self-closing tag 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "What should we do with this ?",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > ignores API mentions 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Hey <@U123456> thanks for letting me know!",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > parses chat extensions 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Heading 1",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Heading 2",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Heading 3",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Heading 4",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Heading 5",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Heading 6",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "This is a paragraph with ",
          "type": "text",
        },
        {
          "marks": [
            {
              "type": "bold",
            },
          ],
          "text": "bold",
          "type": "text",
        },
        {
          "text": " and italics.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Bullet one",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Bullet two",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Number one",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Number two",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "const foo = "bar"",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "Hard breakSoft break",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "HeaderHeaderHeaderCellCellCellCellCellCell",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "And ",
          "type": "text",
        },
        {
          "marks": [
            {
              "attrs": {
                "class": "prose-link",
                "href": "https://linear.app/campsite/issue/CAM-6845/normalize-notes",
                "rel": "noopener noreferrer nofollow",
                "target": "_blank",
                "truncated": false,
              },
              "type": "link",
            },
          ],
          "text": "here",
          "type": "text",
        },
        {
          "text": " is a link. With a ",
          "type": "text",
        },
        {
          "marks": [
            {
              "attrs": {
                "spellcheck": false,
              },
              "type": "code",
            },
          ],
          "text": "inline code",
          "type": "text",
        },
        {
          "text": " mark.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > parses markdown extensions 1`] = `
{
  "content": [
    {
      "attrs": {
        "level": 1,
      },
      "content": [
        {
          "text": "Heading 1",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "attrs": {
        "level": 2,
      },
      "content": [
        {
          "text": "Heading 2",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "attrs": {
        "level": 3,
      },
      "content": [
        {
          "text": "Heading 3",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "attrs": {
        "level": 4,
      },
      "content": [
        {
          "text": "Heading 4",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "attrs": {
        "level": 5,
      },
      "content": [
        {
          "text": "Heading 5",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "attrs": {
        "level": 6,
      },
      "content": [
        {
          "text": "Heading 6",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "content": [
        {
          "text": "This is a paragraph with ",
          "type": "text",
        },
        {
          "marks": [
            {
              "type": "bold",
            },
          ],
          "text": "bold",
          "type": "text",
        },
        {
          "text": " and italics.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "text": "Bullet one",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
        {
          "content": [
            {
              "content": [
                {
                  "text": "Bullet two",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
      ],
      "type": "bulletList",
    },
    {
      "attrs": {
        "start": 1,
        "type": undefined,
      },
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "text": "Number one",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
        {
          "content": [
            {
              "content": [
                {
                  "text": "Number two",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
      ],
      "type": "orderedList",
    },
    {
      "attrs": {
        "language": null,
        "spellcheck": false,
      },
      "content": [
        {
          "text": "const foo = "bar"",
          "type": "text",
        },
      ],
      "type": "codeBlock",
    },
    {
      "content": [
        {
          "text": "Hard break",
          "type": "text",
        },
        {
          "type": "hardBreak",
        },
        {
          "text": "Soft break",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "type": "horizontalRule",
    },
    {
      "content": [
        {
          "marks": [
            {
              "attrs": {
                "class": "prose-link",
                "href": "https://github.com/campsite/campsite/assets/739696/49b398b1-8c03-4255-a759-21b8b53a3f5d",
                "rel": "noopener noreferrer nofollow",
                "target": "_blank",
                "truncated": false,
              },
              "type": "link",
            },
          ],
          "text": "CleanShot 2024-03-22 at 16 42 23@2x",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "HeaderHeaderHeaderCellCellCellCellCellCell",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "content": [
            {
              "text": "Ullamco eiusmod laborum minim nulla adipisicing incididunt occaecat consequat non ipsum ex qui excepteur culpa.",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "blockquote",
    },
    {
      "content": [
        {
          "text": "And ",
          "type": "text",
        },
        {
          "marks": [
            {
              "attrs": {
                "class": "prose-link",
                "href": "https://linear.app/campsite/issue/CAM-6845/normalize-notes",
                "rel": "noopener noreferrer nofollow",
                "target": "_blank",
                "truncated": false,
              },
              "type": "link",
            },
          ],
          "text": "here",
          "type": "text",
        },
        {
          "text": " is a link. With a ",
          "type": "text",
        },
        {
          "marks": [
            {
              "attrs": {
                "spellcheck": false,
              },
              "type": "code",
            },
          ],
          "text": "inline code",
          "type": "text",
        },
        {
          "text": " mark.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > parses note extensions 1`] = `
{
  "content": [
    {
      "attrs": {
        "level": 1,
      },
      "content": [
        {
          "text": "Heading 1",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "attrs": {
        "level": 2,
      },
      "content": [
        {
          "text": "Heading 2",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "attrs": {
        "level": 3,
      },
      "content": [
        {
          "text": "Heading 3",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "attrs": {
        "level": 4,
      },
      "content": [
        {
          "text": "Heading 4",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "attrs": {
        "level": 5,
      },
      "content": [
        {
          "text": "Heading 5",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "attrs": {
        "level": 6,
      },
      "content": [
        {
          "text": "Heading 6",
          "type": "text",
        },
      ],
      "type": "heading",
    },
    {
      "content": [
        {
          "text": "This is a paragraph with ",
          "type": "text",
        },
        {
          "marks": [
            {
              "type": "bold",
            },
          ],
          "text": "bold",
          "type": "text",
        },
        {
          "text": " and italics.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "text": "Bullet one",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
        {
          "content": [
            {
              "content": [
                {
                  "text": "Bullet two",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
      ],
      "type": "bulletList",
    },
    {
      "attrs": {
        "start": 1,
        "type": undefined,
      },
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "text": "Number one",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
        {
          "content": [
            {
              "content": [
                {
                  "text": "Number two",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
      ],
      "type": "orderedList",
    },
    {
      "attrs": {
        "language": null,
        "spellcheck": false,
      },
      "content": [
        {
          "text": "const foo = "bar"",
          "type": "text",
        },
      ],
      "type": "codeBlock",
    },
    {
      "content": [
        {
          "text": "Hard break",
          "type": "text",
        },
        {
          "type": "hardBreak",
        },
        {
          "text": "Soft break",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "type": "horizontalRule",
    },
    {
      "content": [
        {
          "marks": [
            {
              "attrs": {
                "class": "prose-link",
                "href": "https://github.com/campsite/campsite/assets/739696/49b398b1-8c03-4255-a759-21b8b53a3f5d",
                "rel": "noopener noreferrer nofollow",
                "target": "_blank",
                "truncated": false,
              },
              "type": "link",
            },
          ],
          "text": "CleanShot 2024-03-22 at 16 42 23@2x",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "HeaderHeaderHeaderCellCellCellCellCellCell",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "content": [
            {
              "text": "Ullamco eiusmod laborum minim nulla adipisicing incididunt occaecat consequat non ipsum ex qui excepteur culpa.",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "blockquote",
    },
    {
      "content": [
        {
          "text": "And ",
          "type": "text",
        },
        {
          "marks": [
            {
              "attrs": {
                "class": "prose-link",
                "href": "https://linear.app/campsite/issue/CAM-6845/normalize-notes",
                "rel": "noopener noreferrer nofollow",
                "target": "_blank",
                "truncated": false,
              },
              "type": "link",
            },
          ],
          "text": "here",
          "type": "text",
        },
        {
          "text": " is a link. With a ",
          "type": "text",
        },
        {
          "marks": [
            {
              "attrs": {
                "spellcheck": false,
              },
              "type": "code",
            },
          ],
          "text": "inline code",
          "type": "text",
        },
        {
          "text": " mark.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > retains unknown block HTML 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Bad prompt:",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    {
      "content": [
        {
          "text": "<prompt>"Help me with a presentation."</prompt>",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > retains unknown inline HTML 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Bad prompt:",
          "type": "text",
        },
        {
          "type": "hardBreak",
        },
        {
          "text": "<prompt>",
          "type": "text",
        },
        {
          "type": "hardBreak",
        },
        {
          "text": ""Help me with a presentation."",
          "type": "text",
        },
        {
          "type": "hardBreak",
        },
        {
          "text": "</prompt>",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`createMarkdownParser > strips unsupported link attribute values 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "This is ",
          "type": "text",
        },
        {
          "marks": [
            {
              "attrs": {
                "class": "prose-link",
                "href": "https://campsite.com",
                "rel": "noopener noreferrer nofollow",
                "target": "_blank",
                "truncated": false,
              },
              "type": "link",
            },
          ],
          "text": "link",
          "type": "text",
        },
        {
          "text": ".",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
